crates/starknet-os/src/io/output.rs

```diff
@@ -134,99 +134,141 @@ fn next_as_usize<I: Iterator<Item = Felt252>>(output_iter: &mut I, item_name: &s
         .ok_or(SnOsError::CatchAll(format!("{item_name} segment size is too large")))
 }
 
-fn parse_contract_changes<I: Iterator<Item = Felt252>>(output_iter: &mut I) -> Result<ContractChanges, SnOsError> {
+// The reverse of output_contract_state_inner in state/output.cairo
+fn deserialize_contract_state_inner<I: Iterator<Item = Felt252>>(
+    output_iter: &mut I,
+    full_output: Felt252,
+) -> Result<ContractChanges, SnOsError> {
+    let bound =
+        Felt252::from(1u128 << 64).try_into().expect("2**64 should be considered non-zero. Did you change the value?");
+
     let addr = next_or_fail(output_iter, "contract change addr")?;
-    let class_nonce_n_changes = next_or_fail(output_iter, "contract change class_nonce_n_changes")?;
 
-    let two_exp_64 =
-        Felt252::from(1u128 << 64).try_into().expect("2**64 should be considered non-zero. Did you change the value?");
-    let (class_nonce, n_changes) = class_nonce_n_changes.div_rem(&two_exp_64);
-    let (class_updated, nonce) = class_nonce.div_rem(&two_exp_64);
+    let value = next_or_fail(output_iter, "contract change value")?;
+    let (value, n_actual_updates) = value.div_rem(&bound);
+    let (was_class_updated, new_state_nonce) = value.div_rem(&bound);
 
-    let class_hash = if class_updated != Felt252::ZERO {
-        Some(next_or_fail(output_iter, "contract change class_hash")?)
+    #[allow(clippy::collapsible_else_if)] // Mirror the Cairo code as much as possible
```

Comment:
Probably here is a good point to add the link to cairo-lang file. This would be a good reference if there are changes in future versions from cairo-lang

Reply:
Yes I've done that at the top of this function.